### PR TITLE
CE-2966 New Insights list - Pages without infobox

### PR DESCRIPTION
**Pull request by adamkarminski**
/cc: @Wikia/community-engineering 

SQL Review: @wladekb 

`EXPLAIN SELECT tl_from FROM templatelinks WHERE tl_title IN ( 'ModBox', 'Infobox', 'Character', 'TestClassification12', 'Warframe', 'Component', 'Text', 'TestClassification11', 'Planet' );`

`````` +----+-------------+---------------+-------+---------------+---------+---------+------+-------+--------------------------+
| id | select_type | table         | type  | possible_keys | key     | key_len | ref  | rows  | Extra                    |
+----+-------------+---------------+-------+---------------+---------+---------+------+-------+--------------------------+
|  1 | SIMPLE      | templatelinks | index | NULL          | tl_from | 265     | NULL | 12892 | Using where; Using index |
+----+-------------+---------------+-------+---------------+---------+---------+------+-------+--------------------------+```
``````

### CODE REVIEW

File: extensions/wikia/Insights/specials/PagesWithoutInfobox.class.php [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+				} );
+		}
+
+		$sql = ( new WikiaSQL() )
+			->SELECT( 'page_id', 'page_title', 'page_namespace' )
+			->FROM( 'page' )
+			->WHERE( 'page_namespace' )->IN( $wgContentNamespaces );
+
+		if ( !empty( $pagesWithInfobox ) ) {
+			$sql->AND_( 'page_id' )->NOT_IN( $pagesWithInfobox );
```
Comment (lgarczewski) [inline extensions/wikia/Insights/specials/PagesWithoutInfobox.class.php]: This query can have an arbitrary number of pages in it's NOT_IN clause -- may not be wise.

@adamkarminski please do an explain here and ping @Grunny and/or @wladekb

File: extensions/wikia/Insights/models/InsightsPagesWithoutInfoboxModel.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
+<?php
+
+/**
+ * Class InsightsNonportableInfoboxesModel
```
Comment (kamilkoterba) [inline extensions/wikia/Insights/models/InsightsPagesWithoutInfoboxModel.php]: InsightsPagesWithoutInfoboxModel

File: extensions/wikia/Insights/specials/PagesWithoutInfobox.class.php [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+		return $num;
+	}
+
+	/**
+	 * Returns an array with data on pages that do not have an infobox.
+	 *
+	 * @param bool $limit Only for consistency
+	 * @param bool $offset Only for consistency
+	 * @return bool|mixed
```
Comment (kamilkoterba) [inline extensions/wikia/Insights/specials/PagesWithoutInfobox.class.php]: I think this function always return array

File: extensions/wikia/Insights/specials/PagesWithoutInfobox.class.php [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+			->runLoop( $dbr, function ( &$contentPages, $row ) {
+				$contentPages[$row->page_id] = [
+					$this->getName(),
+					$row->page_id,
+					$row->page_namespace,
+					$row->page_title,
+				];
+			} );
+
+		$pagesWithoutInfobox = array_slice( array_diff_key( $contentPages, $pagesWithInfobox ), 0, self::LIMIT );
```
Comment (kamilkoterba) [inline extensions/wikia/Insights/specials/PagesWithoutInfobox.class.php]: What about using $limit here

**Comment by Grunny** [comment]
For the query, assuming templates we classify are always in the template namespace, I'd suggest adding the namespace which improves the query performance:

```
> EXPLAIN SELECT tl_from FROM templatelinks WHERE tl_title IN ( 'ModBox', 'Infobox', 'Character', 'TestClassification12', 'Warframe', 'Component', 'Text', 'TestClassification11', 'Planet' ) AND tl_namespace = 10;
+----+-------------+---------------+-------+---------------+--------------+---------+------+------+--------------------------+
| id | select_type | table         | type  | possible_keys | key          | key_len | ref  | rows | Extra                    |
+----+-------------+---------------+-------+---------------+--------------+---------+------+------+--------------------------+
|  1 | SIMPLE      | templatelinks | range | tl_namespace  | tl_namespace | 261     | NULL |  547 | Using where; Using index |
+----+-------------+---------------+-------+---------------+--------------+---------+------+------+--------------------------+
1 row in set (0.00 sec)
```

**Comment by Grunny** [comment]
Other than query improvement, LGTM.

**Pull request merged**